controllers/nutanixmachine_controller.go

```diff
@@ -327,81 +311,112 @@ func (r *NutanixMachineReconciler) reconcileDelete(rctx *nctx.MachineContext) (r
+	} else {
+		log.V(1).Info(fmt.Sprintf("no task UUID found on VM %s. Starting delete.", *vm.Spec.Name))
+	}
 
-			if rctx.NutanixClientV4 != nil {
-				if err := detachVolumeGroupsFromVM(ctx, rctx.NutanixClientV4, vmUUID); err != nil {
-					errorMsg := fmt.Errorf("failed to detach volume groups from VM %s with UUID %s: %v", vmName, vmUUID, err)
-					conditions.MarkFalse(rctx.NutanixMachine, infrav1.VMProvisionedCondition, infrav1.VolumeGroupDetachFailed, capiv1.ConditionSeverityWarning, errorMsg.Error())
-					log.Error(errorMsg, "failed to detach volume groups", "cluster", rctx.NutanixCluster.Name)
-					return reconcile.Result{}, err
-				}
+	var vgDetachNeeded bool
+	if vm.Spec.Resources != nil && vm.Spec.Resources.DiskList != nil {
+		for _, disk := range vm.Spec.Resources.DiskList {
+			if disk.VolumeGroupReference != nil {
+				vgDetachNeeded = true
+				break
 			}
+		}
+	}
 
-			// Delete the VM since the VM was found (err was nil)
-			deleteTaskUUID, err := DeleteVM(ctx, v3Client, vmName, vmUUID)
-			if err != nil {
-				errorMsg := fmt.Errorf("failed to delete VM %s with UUID %s: %v", vmName, vmUUID, err)
-				conditions.MarkFalse(rctx.NutanixMachine, infrav1.VMProvisionedCondition, infrav1.DeletionFailed, capiv1.ConditionSeverityWarning, errorMsg.Error())
-				log.Error(errorMsg, "failed to delete VM")
-				return reconcile.Result{}, err
-			}
-			log.Info(fmt.Sprintf("Deletion task with UUID %s received for vm %s with UUID %s. Requeueing", deleteTaskUUID, vmName, vmUUID))
-			return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+	if vgDetachNeeded {
+		if err := r.detachVolumeGroups(rctx, vmUUID); err != nil {
+			return reconcile.Result{}, err
 		}
 	}
 
-	// Remove the finalizer from the NutanixMachine object
-	log.Info(fmt.Sprintf("Removing finalizers for VM %s during delete reconciliation", vmName))
-	ctrlutil.RemoveFinalizer(rctx.NutanixMachine, infrav1.NutanixMachineFinalizer)
+	// Delete the VM since the VM was found (err was nil)
+	deleteTaskUUID, err := DeleteVM(ctx, v3Client, vmName, vmUUID)
+	if err != nil {
+		errorMsg := fmt.Errorf("failed to delete VM %s with UUID %s: %v", vmName, vmUUID, err)
+		conditions.MarkFalse(rctx.NutanixMachine, infrav1.VMProvisionedCondition, infrav1.DeletionFailed, capiv1.ConditionSeverityWarning, errorMsg.Error())
+		log.Error(errorMsg, "failed to delete VM")
+		return reconcile.Result{}, err
+	}
+	log.Info(fmt.Sprintf("Deletion task with UUID %s received for vm %s with UUID %s. Requeueing", deleteTaskUUID, vmName, vmUUID))
+	return reconcile.Result{RequeueAfter: 5 * time.Second}, nil
+}
 
-	return reconcile.Result{}, nil
+func (r *NutanixMachineReconciler) detachVolumeGroups(rctx *nctx.MachineContext, vmUUID string) error {
+	createV4Client, err := isPrismCentralV4Compatible(rctx.Context, rctx.NutanixClient)
+	if err != nil {
+		err := fmt.Errorf("error occurred while checking compatibility for Prism Central v4 APIs: %w", err)
+		return err
+	}
+
+	if createV4Client {
```

Comment:
nit: can flatten the function and move the business logic out of the if
```suggestion
	if !canCreateV4Client {
	  // Detaching Volume Groups is only supported with a v4 client.
	  // But this PC version does not support v4 APIs, return early.
	  return nil
	 }
```